test/benchgc/src/benchgc/__main__.py

```diff
@@ -0,0 +1,274 @@
+    # we need to find all args by reading the entry function
+    with open(flags.case, "r") as mlir_file:
+        with gc_mlir.ir.Context() as ctx:
+            module = gc_mlir.ir.Module.parse(mlir_file.read())
+            entry = benchgc.mlir.util.get_entry(module)
+            idx: int = 0
+            # FIXME: only support RankTensorType now
+            for i in entry.type.inputs:
+                args.append(Arg(idx))
+                args[-1].dtype = str(i.element_type)
+                args[-1].shape = list(i.shape)
+                args[-1].set_scalar()
+                idx += 1
+
+            for o in entry.type.results:
+                args.append(Arg(idx))
+                args[-1].dtype = str(o.element_type)
+                args[-1].shape = list(o.shape)
+                args[-1].set_scalar()
+                idx += 1
+elif flags.driver in ["linalg"]:
+    # all arg shape/dt should be provided in single op test
+    for i in range(len(flags.md)):
+        args.append(Arg(i))
+
+    for md in flags.md:
+        colon = md.find(":")
+        if colon == -1:
+            raise Exception("Wrong md format: %s", md)
+        idx = int(md[:colon])
+        args[idx].set_md(md[colon + 1 :])
+
+    from .linalg import mlir_op
+
+    mlir_func = mlir_op[flags.case]
+    module = mlir_func(flags, args)
+else:
+    raise Exception(f"unsupported driver {flags.driver}")
+
+for fill in flags.fill:
+    colon = fill.find(":")
+    if colon == -1:
+        raise Exception("Wrong fill format: %s", fill)
+    idx = int(fill[:colon])
+    args[idx].set_fill(fill[colon + 1 :])
+
+for cmp in flags.cmp:
+    colon = cmp.find(":")
+    if colon == -1:
+        raise Exception("Wrong cmp format: %s", cmp)
+    idx = int(cmp[:colon])
+    args[idx].set_cmp(cmp[colon + 1 :])
+
+entry = benchgc.mlir.util.get_entry(module)
+
+for i, arg in enumerate(args):
+    # use zero filling if the arg is return value
+    set_default_fill(flags, arg, args, i >= len(entry.type.inputs))
+    set_default_compare(flags, arg, args, i >= len(entry.type.inputs))
+
+for arg in args:
+    arg.print_verbose(flags.verbose)
+
+if flags.verbose >= benchgc.util.MODULE_VERBOSE:
+    print(module)
+
+ref_args: List[torch.Tensor] = []
+gc_args: List[torch.Tensor | int] = []
+ref_tensors: Dict[str, torch.Tensor] = {}
+gc_tensors: Dict[str, torch.Tensor] = {}
+
+for i in range(len(args)):
+    tensor = fill_tensor(flags, args[i], i)
+    gc_tensors["%arg" + str(i)] = tensor
+    ref_tensors["%arg" + str(i)] = tensor.clone()
+    ref_args.append(ref_tensors["%arg" + str(i)])
+    if args[i].scalar:
+        gc_args.append(tensor.data_ptr())
+    else:
+        gc_args.append(tensor)
+
+
+# ref_out contains return value of the entry
+ref_out = runner.ref_run(entry, ref_tensors)
+
+# we need to swap the result into the args if some arg is the return value
+if ref_out is not None:
+    for i in range(len(ref_out)):
+        ref_args[0 - i - 1] = ref_out[0 - i - 1]
+
+entry = "entry"
```

Comment:
we also have a variable called `entry` in line 211 with different types, maybe we can change the name of one of them.

Reply:
resolved